File: SpringBootJpa/src/main/java/com/example/java/simple/service/SimpleCustomerService.java
package com.example.java.simple.service;

import com.example.java.simple.domain.Customer;
import com.example.java.simple.repository.CustomerRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@Transactional
public class SimpleCustomerService implements CustomService {

	@Autowired
	private CustomerRepository repository;

	@Override
	public Customer upsert(Customer customer) {
		return repository.save(customer);
	}

	@Override
	public Customer find(Long idx) {
		return repository.findById(idx).orElseThrow(RuntimeException::new);
	}

	@Override
	public List<Customer> findByName(String name) {
		return repository.findByName(name);
	}

	@Override
	public Customer delete(Customer customer) {
		repository.delete(customer);
		return customer;
	}
}

File: SpringBootJpa/src/main/kotlin/com/example/kotlin/simple/service/SimpleCustomerService.kt
package com.example.kotlin.simple.service

import com.example.kotlin.simple.domain.Customer
import com.example.kotlin.simple.repository.CustomerRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.function.Supplier

@Service
@Transactional
class SimpleCustomerService : CustomService {

    @Autowired
    private val repository: CustomerRepository? = null

    override fun upsert(customer: Customer): Customer {
        return repository!!.save(customer)
    }

    override fun find(idx: Long?): Customer {
        return repository!!.findById(idx!!).orElseThrow<RuntimeException> { RuntimeException() }
    }

    override fun findByName(name: String): List<Customer> {
        return repository!!.findByName(name)
    }

    override fun delete(customer: Customer): Customer {
        repository!!.delete(customer)
        return customer
    }
}
